plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java

```diff
@@ -0,0 +1,179 @@
+import org.eclipse.swt.graphics.ImageData;
+import org.eclipse.swt.graphics.ImageLoader;
+import org.eclipse.swt.widgets.Display;
+
+/**
+ * This action provides past image from the clipboard.
+ * 
+ * This action provides past image from the clipboard on Sirius elements that can have workspace image style. This
+ * action create image in image forder of project from clipboard and then set as workspace image on all selected nodes.
+ * 
+ * @author Séraphin Costa
+ */
+public class PasteImageAction extends Action {
+
+    // There are no clipboard listener in SWT, so we need to use the AWT clipboard listener.
+    java.awt.datatransfer.Clipboard awtClipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
+
+    class ClipboardListener implements FlavorListener {
+        @Override
+        public void flavorsChanged(FlavorEvent e) {
+            // This is an AWT event, not an SWT event,
+            // so we are not in the thread UI and we need to execute this code in thread UI.
+            Display.getDefault().syncExec(() -> {
+                updateState();
+            });
+        }
+    }
+
+    /**
+     * Default constructor.
+     */
+    public PasteImageAction() {
+        super();
+
+        setId(ActionIds.PASTE_IMAGE);
+        setText(Messages.PasteImageAction_text);
+        setToolTipText(Messages.PasteImageAction_toolTipText);
+        setAccelerator(SWT.CTRL | SWT.ALT | 'V');
+
+        setImageDescriptor(DiagramUIPlugin.Implementation.getBundledImageDescriptor(DiagramImagesPath.PASTE_IMAGE_ICON));
+
+        updateState();
+
+        // listen clipboard for update state
+        awtClipboard.addFlavorListener(new ClipboardListener());
+    }
+
+    /**
+     * Update the state of this action.
+     * 
+     * Updates the active state of the action according to the environment.
+     * 
+     * This method must be executed in the UI thread.
+     */
+    public void updateState() {
+        setEnabled(canBeEnabled());
+    }
+    
+    /**
+     * {@inheritDoc}
+     * 
+     * @see #org.eclipse.sirius.diagram.ui.tools.internal.actions.layout.PasteImageAction PasteImageAction class
+     *      documentation
+     */
+    @Override
+    public void run() {
+        List<BasicLabelStyle> selectedStyles = SetStyleToWorkspaceImageAction.getStyles();
+        ImageData imgData = fetchImageFromClipboard();
+
+        // we check if the clipboard contains image and we have at least one element selected
+        if (imgData != null && !selectedStyles.isEmpty()) {
+            // get the first style to provide the context for file creation and domain for command
+            BasicLabelStyle firstStyle = selectedStyles.stream().findFirst().orElseThrow();
+            TransactionalEditingDomain domain = TransactionUtil.getEditingDomain(firstStyle);
+
+            if (domain != null) {
+                // prepare image file creation
+                String filename = ImageManager.generateName("png"); //$NON-NLS-1$
```

Comment:
That means the image will always have png extension?

Reply:
Yes, I always save the image in PNG, see https://github.com/scosta-obeo/sirius-desktop/blob/9766cb963a29a1693c940669fc3f5d947dd91e43/plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/layout/PasteImageAction.java#L162